### ISSUE TITLE
Smart tabs

### ISSUE THREAD

**Issue by rpanak** [editorconfig/editorconfig#323]
I suggest a `smart_tabs` property should be added. IntelliJ IDEA, Vim and Emacs support the feature, see [Vim wiki](http://vim.wikia.com/wiki/Indent_with_tabs,_align_with_spaces), [help for IDEA](https://www.jetbrains.com/help/idea/2017.1/code-style-java.html). The idea behind the property is pretty simple: indent with tabs, align with spaces [explanation] (http://www.lb-stuff.com/tabs). Two possible values: `true` and `false`. The setting is quite standard in my opinion so there's no reason (hopefully) for not including it.

**Comment by rpanak** [comment]
Or better yet, let's add a new valid value (`smart`) to `indent_style` .

**Comment by xuhdev** [comment]
It looks like the definition of `smarttab` is consistent across all editors and it is also well supported. @editorconfig/board-member Do you guys wanna have a vote?

**Comment by jedmao** [comment]
That would be an excellent addition to EditorConfig. You have my vote!

**Comment by florianb** [comment]
🤔 Do i miss any proof that smart tabs are supported "across all editors"? And what is this vote asking for, for adding a smart_tabs property to the standard yet or starting a process to define a a smart_tabs property?

**Comment by johan** [comment]
I second the lack of clarity about what a vote means.

It would probably go a long way writing a concise, precise definition of what the `smart_tabs` property does, instead of linking to a bunch of posts describing it, or arguing what some author thinks is right, so votes refer to the same thing. I suggest voting happens on a pull request to the spec, detailing what the definition of the property would be.

An attempt at a sketch, that I'd hope someone who wants this can improve upon to turn into that:

`smart_tabs`: when present, either `true` or `false`. `true` makes all code indentation use hard tabs, while comment alignment uses spaces. (If `true`, `tab_width` also needs to be set in `.editorconfig` for the combination to work, as a combination of tabs and spaces aligns randomly across editors without a well-defined tab width.)

I'd lean towards going as far as specifying this setting to only take effect when both are set, but with the poor feedback from editors about why they behave the way they do (or don't behave the way they don't), and cascading `.editorconfig` files (setting one property at one level and the other higher up in the directory structure), and how hard it is for a human to intuit whether both are set or not, that's probably not a great idea in practice.

**Comment by hgraeber** [comment]
I have checked QtCreator and found it supports smart tabs alittle bit, but not completly. Adding this functionality with my plugin is possible, but extending QtCreator would be better. 

What I have discovered while thinking about it is that there are some dependencies between options. Setting`smart_tabs = true` only makes sense together with `indent_style = tabs` and maybe `indent_size = tab`.

The consequence is, that instead of having a new boolean `smart_tab` property it would be better to have a new alternative value for `indent_style`, named `smart_tabs` or simply `smart`.

**Comment by jedmao** [comment]
@johan it shouldn't matter what your `tab_width` setting is, because as long as you stick with tabs for indentation and spaces for alignment, you can have any `tab_width` you want without breaking the visual alignment of your code (assuming the editors are doing it properly). See the following illustration:

![tabs-vs-spaces](https://cloud.githubusercontent.com/assets/1058243/18036844/49a9426c-6d3b-11e6-98ae-2c9c3db623ae.gif)

Here is what I am imagining this rule would do:

## `smart_tabs = true`
Enables tabs for indentation and spaces for alignment in editors that support the feature.
- For editors that do _NOT_ support the feature, `indent_style = tabs` should be implied. The responsibility remains upon the human to enforce the following:
  - Tabs for indentation _only_ and _never_ for alignment.
  - Spaces for alignment _only_ and _never_ for indentation.
- **Warning: this setting overrides the `indent_style` setting to the value of `tabs`.**
  - `indent_style = tabs`

**Comment by hgraeber** [comment]
@jedmao 

> Warning: this setting overrides the indent_style setting to the value of tabs.
>    - indent_style = tabs

That's the reason I would suggest to use `indent_style = smart_tabs` instead of a new `smart_tabs` property.

**Comment by jedmao** [comment]
@hgraeber I can't think of a con to your suggestion; though, would there be any logical reason for anyone to set `indent_style = smart_tabs` over just `indent_style = tabs`?

What I'm asking is what if we just assumed all people who have `indent_style = tabs` want smart tabs? Is that not a safe assumption?

**Comment by xuhdev** [comment]
My suggestion is to vote for adding an `smarttab` option to `indent_style`. @jedmao The only difference is the handling of alignment. When set to `tabs`, it would mean to align using a mixture of tabs and spaces. When set to `smarttab`, it would mean to align using spaces. That's my understanding.

**Comment by jedmao** [comment]
My point is that there is no good reason for anyone ever to align with tabs, so why would we enable or encourage people to think that's a good idea? It's a fundamentally wrong idea and it's one of the biggest reasons many people prefer spaces over tabs for indentation, because so many people have abused tabs, not knowing that they are _never_ supposed to be used for alignment.

**Comment by rpanak** [comment]
@florianb Nobody stated that smart tabs are supported across all editors. It's worth pointing out that it's definitely not supported by all editors. But I still think it's a worthy addition to the spec.

@johan It would probably go a long way having information in the faq that you expect an issue to contain a spec right from the beginning, instead of complaining in comments :)

Anyway, unaware of your process I just wanted to start the discussion and I'm pleased to see that there is a will to work on this. I agree that we first need a solid definition.

A couple of remarks towards the spec sketch from @johan:

- there is no need for `tab_width` to be set, if smart tabs are used. The idea is that everyone can have his own `tab_width` setting (depending on how much visual indentation he needs), because only tabs are used for indentation and only spaces for alignment.
- there is, however, a limitation for the combination of values of `tab_width` and `indent_size`. If `indent_size` is not a multiple of `tab_width`, you'll need a combination of tabs and spaces to indent according to the setting, thus no smart tabs.
- alignment with smart tabs uses spaces for all alignment, not just comments
- there are more combinations that don't make sense and should be at least documented, e.g. `smart_tabs` set to `true` with `indent_style` set to `space` 

Therefore, I'd rather add `smart_tab` as another possible value to `indent_style`:

-----
### `indent_style`

Indentation Style

_The values are case insensitive. They will be lowercased by the core library._

**Possible Values**

- `tab`
- `space`
- `smart_tab`

**_NOTE:_** `smart_tab` uses TABs for indentation and SPACEs for alignment, so setting `indent_style` to `smart_tab` should be ignored unless `indent_size` is set to be a multiple of `tab_width`.

**Supported By**

- all plugins (`smart_tab` value is not supported by all plugins)

-----

Another option is leaving `indent_style` as it is and introducing a property `align_style` with the same set of values. This would be in my opinion the best description of reality. Current description of `indent_style` is incorrect because it is currently used in editors both for indentation as well as alignment. However, it is not a pragmatic description. This _error_ of mixing and combining alignment and indentation is the status quo in most (if not all) editors and it's unreasonable to think all of the editor teams will correct it any time soon.

Happy to hear your opinion.

**Comment by rpanak** [comment]
@jedmao I second your opinion, smart tabs make the most sense, while just tabs don't. However, there are many religious views connected to this issue and, in my view, EditorConfig describes the current situation with editors (more or less) and they allow this setting. Therefore, we should keep tabs as a valid option.

@xuhdev It's not that easy, unfortunately. For some strange reason you might want to indent by 5 characters each level with tab size set to 4 characters and indent style set to tabs. You can achieve this only with a combination of tabs and spaces for indentation.

**Comment by hgraeber** [comment]
Thinking further I see one problem with extending indent_style: Compatibility with older implementations of the editorconfig spec. Those do not know the value `smart_tab` and even don't know to fallback to `tab`. Adding a new property instead, like originally suggested, does not have this problem. It will only be ignored.

**Comment by rpanak** [comment]
@hgraeber Is there a need for fallback? For the old implementations `smart_tab` should be the same as setting `indent_style` to any other unsupported value.

**Comment by jedmao** [comment]
@hgraeber is correct. It essentially works as `unset`, unsetting any `indent_style` settings defined in any parent directories. A completely new setting would fix that problem and be the most backwards compatible.

@rpanak, indenting by 5 spaces with tab size set to 4? That doesn't even make any sense to me. How would the `.editorconfig` configuration look like? Strangeness aside, I'm not aware of any religious views supporting the mixture of tabs and spaces for alignment. I think everyone can agree that's a bad idea. The confusion is that some people hear "mixed tabs and spaces" and assume they are being mixed for alignment, which is the root of the problem. If, however, tabs are used for indentation only and spaces for any alignment that follows, there will **_NEVER_** be an issue with alignment, no matter what setting you have for tab width.

---

No matter how many times it's stated or even [the animated GIF above](https://github.com/editorconfig/editorconfig/issues/323#issuecomment-327815131), sometimes I feel like people just aren't "getting it." If you "get it" then there is no reason to fret about what setting you have for tab width. Any setting is perfectly acceptable and does not break the alignment.

<img src="https://media.giphy.com/media/l0IylSajlbPRFxH8Y/giphy.gif">

## Conclusion
I don't want to perpetuate the confusions above and avert people from using tabs for indentation and spaces for alignment, which is perfectly the right way to do hard tabs.

### Indentation
- Option 1 – Use tabs only, never spaces.
  - Users are free to set tab width to their own preference w/o breaking alignment, assuming alignment uses spaces only.
- Option 2 – Use spaces only, never tabs.
  - Tab width is irrelevant and has no affect.
  - **_Warning: You are forcing everybody to see the same tab width._**
  - **_Warning: Arrow-key navigation will stop on every space. Depending on the editor, delete and backspace could also potentially stop on every space._**

### Alignment
Follows indentation. Use spaces only, never tabs. No exceptions, regardless of indentation setting.

**Comment by rpanak** [comment]
@jedmao 

1. I meant there are religious views about indent_style tab. Or maybe not. Hard for me to know as most of those indent style discussions I've been part of end up as a holy war mostly coming down to personal habit than to actual arguments. But then again, those "religious" views are seldom based on arguments :)
2. The 5 and 4 values are odd on purpose to illustrate the problem. The editors allow such values. Period. And to some degree EditorConfig acknowledges that, as it separates indent_size from tab_width. For me .editorconfig is a settings store shared across multiple editors, not a best practice guide. It's hard for me to be an advocate for the other side (as I agree with you on what the best option is), but I'd still give them the freedom of choice, since the editors allow it.
3. With your proposal going through all the editors not supporting smart tabs would become incompatible.
4. As for backwards compatibility of .editorconfig. You're the experts I guess.

For me it boils down to two important questions:
a) do you want to mimic what editors do or do you want to pave a new way for a better style?
b) do you want to implement the changes in an easier to understand/document way (less properties influencing each other) or to be backwards compatible with older .editorconfigs?

**Comment by jedmao** [comment]
@rpanak those are all excellent concerns and questions.

1. There are definitely religious views about whether to use tabs or spaces for indentation. There are actually good arguments on both sides too.
1. Are you saying that editors allow you to have a separate indent size vs. tab size? Maybe some editors do, but that doesn't mean they should. [Visual Studio Code](https://code.visualstudio.com/), for example, does not support this horrible feature and I would imagine that the lack of support was very intentional and a good idea on their part.
1. I agree that my proposal didn't account for backwards compatibility. That is definitely a concern and I can't imagine it would be a good idea to introduce a breaking change.
1. The best backwards-compatible way to support this feature is, again, to introduce a completely new setting of `smart_tabs = true`.

I do want to use built-in editor features whenever possible. If an editor has philosophically chosen to go a certain direction, I do want to honor that in the respective EditorConfig plugins, so long as it doesn't break away from the [plugin guidelines](https://github.com/editorconfig/editorconfig/wiki/Plugin-Guidelines). In fact, that was really one of the main objectives of writing those very guidelines.

I don't think it's EditorConfig's place to force controversial conventions on people, but that's why I'm arguing that nobody really supports mixing tabs with spaces. That part is not one of the controversial parts of the holy wars of tabs vs. spaces.

To answer your last question, I want to be backwards compatible.

**Comment by rpanak** [comment]
@jedmao 
2. Yep, they do allow it (as I was trying to point out, even Editor Config has two separate properties). All the IDEs from JetBrains, AndroidStudio (the main Android IDE, still based on JetBrains's IndelliJ IDEA), Vim, Visual Studio. I couldn't find it for VSCode, Sublime or Emacs, but I'm not using those that often. I guess there are more. Don't know the reason, but maybe they have some.

So to sum up, going forward it seems best to introduce new `smart_tabs` property. The interactions between other properties should be documented carefully.
The feature will be only turned on if:

1. `indent_style` is set to `tab`
2. `indent_size` is a multiple of `tab_width`
3. `smart_tabs` is set to `true`

**Comment by jedmao** [comment]
@rpanak it sounds like the older IDEs are the ones that support both settings at the same time and the newer ones do not. Maybe there's a reason for THAT!

1. Why not override `indent_style` with the value of `tab` if `smart_tabs = true`? Couldn't it be implied?
1. What does `indent_size` is a multiple of `tab_width` mean? `tab_width` is commonly set to a value of `4`, so a multiple would be `8` or `12`? Makes no sense.

**Comment by rpanak** [comment]
@jedmao 
1. I guess that's possible. The other members should voice their opinion about this, too.
2. Yes, that's what it means. Once you allow both settings, it makes perfect sense. If it's not a multiple, you can't indent each level by `indent_size` with tab characters only. You'd have to use spaces.

**Comment by johan** [comment]
@jedmao / @rpanak: I don't see how `smart_tabs` can support alignment across indentation levels without making `tab_width` well-defined. It seems to break trivially, i e (`$TAB` signifying a hard tab):

```
$TABvar foo        = 'FOO',          // this aligned comment
$TAB    barBazQuux = 'BAR-BAZ-QUUX'; // section will not
$TABif (gazonk) {                    // remain aligned across
$TAB$TABvar oops  = 'OOPS';          // different tab widths
$TAB}                                // (this one works at 4)
```

With a tab_size of 8 (a typical default different from above), that comes out like this, instead:

```
        var foo        = 'FOO',          // this aligned comment
            barBazQuux = 'BAR-BAZ-QUUX'; // section will not
        if (gazonk) {                    // remain aligned across
                var oops  = 'OOPS';          // different tab widths
        }                                // (this one works at 4)
```

> @johan It would probably go a long way having information in the faq that you expect an issue to contain a spec right from the beginning, instead of complaining in comments :)

Principle of charity, please – we are all trying to be constructive (I hope). This sounds like a good idea. I'll open a separate issue for that discussion.

**Comment by hgraeber** [comment]
@jedmao Overriding `indent_style` with the value `tab` of if `smart_tabs = true` doesn't make sense compatibility wise. Older editorconfig implementations need the latter explicitly as a fallback. Instead having `indent_style` not set to `tab` together with `smart_tabs = true` is some sort of error. Programs that have some means of giving a not too annoying message should report some warning for that. The same holds for `indent_size = tab`.

So it should work as follows:

    tab_with = 4        ; Optional, smart_tabs work with any tab width, thats their purpose ;-)

    indent_style = tab  ; These provide a proper fallback for older editorconfig implementations,
    indent_size = tab   ;     that do not known smart_tabs, but alignment must be done by the user

    smart_tabs = true   ; Enable smart tabs, the two leading indent_ settings are mandatory and are
                        ;    enforced, but should give a warning if they are missing until the smart_tabs property
                        ;    has spread over all implementations.

Nice example of smart tabs... ;-)

**Comment by florianb** [comment]
I would like to refer to the definition of ["elastic tabstops"](http://nickgravgaard.com/elastic-tabstops/). (It took me some time to find the site again - so sorry for the delay).

Elastic tabstops seem to be a mature concept. And they're not using spaces for alignment.

So i'd lean towards using `elastic_tabstops` (instead of smart_tabs), combined with an explicit `indent_style = tab` (for backwards compatibility).

In my opinion, `indent_size` (in combination with tabs) is for custom layouting purposes only -- and that's how i am using the combination of style and size in my projects.

**Comment by hgraeber** [comment]
@florianb I think this is OT here. Elastic tab stops are nice and work even with proportional fonts. But they definitely need proper support from the editor or viewer and may don't have this. Our goal is to align more simple setting common by most editors.

Smart tabs are much simpler and can even be done by hand for editors that do not support them. The `smart_tabs` property is only meant for letting the editor help you with that.

**Comment by florianb** [comment]
Oh i just saw there is #317 declining elastic tabstops.

Thanks for explanation @hgraeber - i currently just don't see that "smart-tabs" are wider supported than elastic tabstops. Don't get me wrong but earlier attempts to add new functionality to EditorConfig were declined to lacking support of editors. Now we're discussing a feature which as wide spread as elastic tabs and it seems to be even harder to implement.

Harder to implement?

It is not possible to detect "alignment" spaces in code without knowldege about the semantics, isn't it? How to determine the right amount of spaces to choose when pressing the first tab?

I can't second the argument this is something you can do by hand if your editor won't support it. Adjusting spaces by hand is disgusting. That's what tabstops were invented for.

Using interline spaces over tabstops may force developers to handle spaces by hand (to produce well formed files), not correctly formatting tabstops may only lead to a distraction.

**Comment by hgraeber** [comment]
@florian Don't understand me wrong, I like elastic tabs. But an editor or viewer must have support not only for editing, but - more important - for viewing, too. And many editors or viewers have not.

Smart tabs need no special support for viewing and can - if neccessary - be be entered without editor help. But with some help by the editor it's easier and more consistent between multiple users. That's what `smart_tabs` property comes in.

If you want elastic tabs, then this should discussed at #317. Let's pick the lower hanging fruits here.

**Comment by florianb** [comment]
So i guess i really don't get what this "standardization attempt" aims for. I thought it would be about giving developers the freedom to ensure their development tools enable support for code alignment (see f.e. [this](https://shkspr.mobi/blog/2014/11/why-i-vertically-align-my-code-and-you-should-too/) blogpost) in general. But i guess i got it now, you all want just smart tabs being activated.

I feel it's a disadvantage that we don't have unified goal for the Editorconfig-standard.

Because either we want people to configure their possible editor-features or we want to maintain a mature superset of features where it is pretty likely that it can be supported in every editor.

I have the strong opinion that this would be a domain-specific property since smart-tabs would need semantic knowledge about the language. And i don't see this being a "low" hanging fruit - implementing this carefully will cause some efforts.

**Comment by jedmao** [comment]
@johan I like your example of the aligned comment section not aligning with different tab widths. I never considered that scenario before.

I've seen elastic tab stops before. Very awesome, just lacks editor support, same as smart tabs.

I think it's OK if all editors don't support a feature though. We should leverage built-in editor features where possible and if they don't support it, then the plugin doesn't support that particular setting. As plugin authors, we shouldn't feel obligated to support an EditorConfig setting, just because it's specified; rather, we should aim to support settings where possible and not fret about 100% support.

That's my $0.02.

**Comment by hgraeber** [comment]
@johan 

> I've seen elastic tab stops before. Very awesome, just lacks editor support, same as smart tabs.

Contrary to elastic tabs, smart tabs need no special support when viewing files and editing may be supported by editors, but it can be done by hand for dumb editors.

Redmine for example has no editorconfig support yet and fixed tabs of 8. With smart tabs even redmine will show files somewhat right. Something impossible with elastic tabs, which needs support for editing AND viewing files.

Nevertheless elastic tabs are interesting and support for them may be added, too -> #317. Would be great every editor supports them and one can use proportional fonts for coding!

**Comment by rpanak** [comment]
@johan Haven't considered that example, thanks for that. I guess I've seen very little code using it myself, so I've been oblivious to this style (not suggesting it's not being used). Note that I'm not advocating for everyone to use smart_tabs. It's my personal preference and as your example shows, it's no silver bullet. However, it's a valid option and EditorConfig should provide it.


@hgraeber  
> `indent_size = tab`

As I already stated before, the limitation should be that `indent_size` is a multiple of `tab_width`. Your proposal implies that limitation, but it's unnecessarily strict.

**Comment by tilmanpotthof** [comment]
Another valid use case for alignment with spaces is api documentation in multiline comments:

         /*
          * My comment
          * 
          * @param myParam
          */
         public myFunction() {...

**Comment by jedmao** [comment]
@tilmanpotthof, that's not a valid use case. Lines 2-5 would use a tab followed by one space for alignment.

**Comment by xuhdev** [comment]
For vote, what I meant was to whether we should add a `smarttab` value to `indent_style`. I haven't followed through the discussion, unfortunately, due to my severe time constraints in the past few days. If any of you think you can initiate a vote more clearly than I do at this stage, feel free to do so.

**Comment by jedmao** [comment]
@xuhdev to save you the time reading, I think the issue with that approach is mainly that it's not backwards compatible (i.e., it will behave as an `unset` for plugins that haven't yet been updated to understand `indent_style`).

**Comment by jgeerds** [comment]
Backward compatibility is a general problem and the only way to handle this properly is to use version numbers (e.g. `version = 1` inside `.editorconfig`). Let's call the current file format version `1`. This version is also implied if the `.editorconfig` doesn't contain a version number.

Then we can define a new version where `smart_tab` is a valid value for `indent_style` and give plugin developers time to support this new format. Then, in year 2020 or something like that, we enforce the compatibility of the new file format. This would be the first time developers can actually use `version = 2` inside their config files.

…and because that's take too much time, we could can now introduce smarttabs in a backward compatible way (e.g. a new field called `smart_tabs` or something like that). At the same time we deprecate the use of the `smart_tabs` property so that plugin developers know, that this property disappears in the next version.

Sounds difficult but I think this is the only way to extend editorconfig while also maintain a clear configuration syntax.

**Comment by jedmao** [comment]
@jgeerds that's an interesting proposition. I think what you're saying is that we add this to the spec right now:

```ini
root = true
version = 1.0

[*]
indent_style = space
indent_size = 2
```
And in 2020 or whenever we feel confident that all the plugins are supporting versioning, then we introduce version 2 like:
```ini
root = true
version = 2.0

[*]
indent_style = smarttab
```
Totally backwards compatible. I like it!

The only action item for now is that we add versioning to the spec, yeah? Important to note that the version relates to the specification and not the actual version of the EditorConfig parser.

We can start a roadmap document to convey our future intentions too.

@xuhdev I'm a bit concerned that people might typo `smarttab` as `smartab`, so what do you think if we support both as a precaution?

**Comment by xuhdev** [comment]
If backward compatibility is an issue, I personally prefer to create a new property. This can avoid us introducing version numbers in the format itself. For example,

```
indent_size = tab
smarttab = true
```

Current plugins just ignore the smarttab option and still work mostly decently according to what is expected. Note that, in the future, it's also impossible for all editors to support this feature, due to the limit of the editors themselves. If we introduce this feature in this way, those editor plugins can also avoid updates.

@jedmao I think we should stick to one to avoid confusion among plugin developers (`smarttab`)...

**Comment by jedmao** [comment]
@xuhdev then you get into the issue of what do we do with the following?
```ini
indent_size = space
smarttab = true
```
There has to be a priority, right? Then it's just confusing.

**Comment by xuhdev** [comment]
@jedmao In my mind, if `indent_size` is space, then no matter what happens, we use space for indentation. We can think about it in this way: `indent_size` specifies whether to use `tab` as indentation character, and `smarttab` specifies whether the tab indentation should be smart. What do you think?

**Comment by jgeerds** [comment]
@jedmao: I don't like the idea in supporting both versions (**smarttab** and **smartab**). This might be user friendly but on the other hand also confusing. Alternative: **smart_tab** (underscore)

My issue with a new`smarttab = true` property is:
- It doesn't scale very well. Think of supporting elastic tabsstops (#317) in further versions. Should we also implement this as `elastic_tabstop = true` ? This could lead to "invalid" configurations in combination with other settings (for example with `smarttab = true`). Then we have to carefully document how plugins should handle those invalid configurations.
- It clearly belongs to an already existing configuration option: **indent_style**


Instead of versioning our config format we could — at least in this case — also define a new policy: Plugins should not only handle new config properties but also new values for existing properties (e.g `smart_tab` or `elastic_tabstop`). Notify all plugin developers about this new policy and give them time until end of the year. Finally we update our config specs next year.

**Comment by jedmao** [comment]
@jgeerds that might work with some projects, but it's pretty difficult with this one because some of the plugins aren't regularly maintained (e.g., the Visual Studio extension). The glory of the backwards compatible solution is that we don't have to present an ultimatum (a deadline), which works quite well for this project.

@xuhdev I think what you're saying is that `smarttab = true` would be completely ignored in [my example above](https://github.com/editorconfig/editorconfig/issues/323#issuecomment-331223235). That's fine, but can we just stick with `smart_tabs` as the OP suggested? Reasons why:

1. It's consistent with current implementations in both [VIM](http://vim.wikia.com/wiki/Indent_with_tabs,_align_with_spaces) and [IDEA](https://www.jetbrains.com/help/idea/2017.1/code-style-java.html) as [mentioned by the OP](https://github.com/editorconfig/editorconfig/issues/323#issue-253024542).
1. It solves [the typo issue](https://github.com/editorconfig/editorconfig/issues/323#issuecomment-331212878) with the `_` delimiter.
1. It's consistent with other EC rules that are also snake case.

**Comment by xuhdev** [comment]
@jedmao I agree with you on using `smart_tab` instead of `smarttab`. Should we make a voting thread?

**Comment by jedmao** [comment]
@xuhdev, yes and to be clear, `smart_tabs` was the original suggestion (in the plural). This is also what VIM and IDEA call it (i.e., "Smart Tabs").

**Comment by jgeerds** [comment]
Yep, `smart_tabs` sounds good to me.

**Comment by xuhdev** [comment]
OK, we can do `smart_tabs`.

**Comment by xuhdev** [comment]
The voting is available at editorconfig/editorconfig-vote#1

**Comment by Mpdreamz** [comment]
I voted 1 on the voting issue. despite being personally on board with the tabs for indenting, spaces for alignment idiom.

Could this not be an editor specific extension property? Quick read of the VIM spec seems like a plugin is needed there too. This property seems like a very hard/impossible thing to try an implement in all editors. 

I know not all properties are mandatory to be implemented in all editors but this property in my opinion complicates an already confusing part of the spec for the benefit of a few.

**Comment by jedmao** [comment]
@johan can you elaborate on what [damage](https://github.com/editorconfig/editorconfig-vote/issues/1#issuecomment-331369510) will be done by such a setting? I don't think we expect plugin developers to actually build the feature; just, to enable it if available (i.e., the editor supports it natively). I didn't realize VIM required a separate plugin be installed for the feature.

Is the damage done because it's so terribly difficult to implement that the implementations may not work consistently all the time; thus, introduce imperfections in the indentation? Also, complex as it is, different plugin developers will have totally different implementations, making it even more inconsistent between editors?

If that's the case, I'll be happy to change my vote from a 2 to a 1. Seems way more high risk than originally perceived.

**Comment by xuhdev** [comment]
The voting results of editorconfig/editorconfig-vote#1 have decided to support smart tab "by creating a smart_tabs property, with two options true and false".

**Comment by johan** [comment]
@jedmao The damages aren’t huge; just editorconfig having a hand in indentation drift between editors for the under-specified case with comments across indent levels I noted earlier (where an arbitrary user-specific default tab width will affect outcomes, unless the spec nails down that the setting only gets enabled if editorconfig also nailed down the tab width for `smart_tabs` enabled files), which feels a bit careless.

In other words: a smell, not a great risk. Whether with that amendment or not, there is certainly a bit of complexity for plugin and editors to implement to fulfill the promise of consistency.

**Comment by ArmanHayots** [comment]
So I can already put `smart_tabs = true` for the future use?

**Comment by jedmao** [comment]
@ArmanHayots yes.

**Comment by hax** [comment]
What is the precise semantic of this new `smart_tabs` property? The thread is too long to catch.

**Comment by xuhdev** [comment]
@hax No, not really. It is in doubt that how many editors can support this feature.

**Comment by Naheulf** [comment]
I don't understand one thing: why name it "smart_**tab**" if it mean use **spaces**.

**Comment by RdeWilde** [comment]
Please implement.

**Comment by tysonclugg** [comment]
Please excuse me for dredging up an old thread, but I'd like to confirm that the final behaviour of `smart_tabs` will ensure the following bash script won't be munged (particulalry the "use multiple times to increase verbosity" line):
```bash
#!/bin/bash

show_help() {
	cat<<-EOD
	Options:
	  -h, --help     show this help
	  -v, --verbose  verbose output
	                 (use multiple times to increase verbosity)
	EOD
	exit 0
}

[ $# -eq 0 ] && show_help
# TODO: everything else...
```

The significance of tabs vs spaces in the above script is very important, since `<<-` style here documents have tabs stripped (but not spaces), and the spaces are being used to align the help output.

If `smart_tabs` is not the right feature for this, I'd love to know what is.  Thanks!

**Comment by darkstego** [comment]
@tysonclugg 

This depends on the implementation. Emacs for example doesn't touch anything in <<- block when you ask it to indent the file. Even if smart-tabs is off it will indent the first line with spaces and keep the remaining untouched.

Smart Tabs does not break this as these lines should not be indented, but the alignment has to be managed manually.

**Comment by jycr** [comment]
Is there any news about finalizing this feature?

**Comment by SethFalco** [comment]
Hmm, I can see there was a vote to add the `smart_tabs` property to the spec, but then no such update was made to the spec. 🤔 

https://github.com/editorconfig/editorconfig-vote/issues/1

Was this a mistake?

**Comment by xuhdev** [comment]
@SethFalco Strange, seems we have totally forgotten it, probably because the spec wasn't formally organized at that time. It's not difficult to add it to the spec, however, we likely lack contributors to get it implemented in various plugins.  Let me open a PR to add it to the spec and get some feedback there.

**Comment by cowwoc** [comment]
@xuhdev Hello from 2023. Where is the PR you were referring to? :)

**Comment by florianb** [comment]
Hey @xuhdev - i came by to just open the PR and read through the thread again.

I conclude:

1. We want to allow smart indentation (`smart_tabs` et al.) in general.
2. We want it to be backward compatible.

To make this discussion a bit more productive i took the liberty of opening a pr where i (in opposition of the formerly announced `smart_tabs` property) propose the `indent_method` property to avoid future discussions about adding another indentation mechanisms.

This isn't set into stone - so if that's against all opinions we're free to revert it to the `smart_tabs`proposal.

Please let's move the final discussion to the PR: https://github.com/editorconfig/specification/pull/38

**Comment by xuhdev** [comment]
@florianb Thank you for taking this on -- appreciate it, will continue discussion there
